presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveStorageFormats.java

```diff
@@ -67,6 +67,138 @@
+                        "1967-01-02 12:01:00.1114", // hundreds of micros, rounds down in millis
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.111",
+                        "1967-01-02 12:01:00.111400",
+                        "1967-01-02 12:01:00.111400000"),
+                timestampAndPrecision(
+                        "1967-01-02 12:01:00.1115", // hundreds of micros, rounds up in millis (smallest)
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.112",
+                        "1967-01-02 12:01:00.111500",
+                        "1967-01-02 12:01:00.111500000"),
+                timestampAndPrecision(
+                        "1967-01-02 12:01:00.111499", // micros, rounds down (largest)
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.111",
+                        "1967-01-02 12:01:00.111499",
+                        "1967-01-02 12:01:00.111499000"),
+                timestampAndPrecision(
+                        "1967-01-02 12:01:00.1113334", // hundreds of nanos, rounds down
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.111",
+                        "1967-01-02 12:01:00.111333",
+                        "1967-01-02 12:01:00.111333400"),
+                timestampAndPrecision(
+                        "1967-01-02 12:01:00.1110019", // hundreds of nanos, rounds down in millis and up in micros
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.111",
+                        "1967-01-02 12:01:00.111002",
+                        "1967-01-02 12:01:00.111001900"),
+                timestampAndPrecision(
+                        "1967-01-02 12:01:00.111901001", // nanos, rounds up in millis and down in micros
+                        NANOSECONDS,
+                        "1967-01-02 12:01:00.112",
+                        "1967-01-02 12:01:00.111901",
+                        "1967-01-02 12:01:00.111901001"),
+                timestampAndPrecision(
+                        "1967-01-02 23:59:59.999999999", // nanos, rounds up to next day
+                        NANOSECONDS,
+                        "1967-01-03 00:00:00.000",
+                        "1967-01-03 00:00:00.000000",
+                        "1967-01-02 23:59:59.999999999"),
+                timestampAndPrecision(
+                        "1967-12-31 23:59:59.999999499", // nanos, rounds micros down (largest), rounds millis up to next year
+                        NANOSECONDS,
+                        "1968-01-01 00:00:00.000",
+                        "1967-12-31 23:59:59.999999",
+                        "1967-12-31 23:59:59.999999499"));
+
+        // Make post-epoch tests
+        Stream<TimestampAndPrecision> postEpoch = preEpoch.stream()
+                // Exclude years outside of 1960-1968
+                .filter(timePrec -> !timePrec.getWriteValue().matches("^196[^9]"))
```

Comment:
This makes it not so explicit about what's actually tested.
Let's enumerate the test cases explicitly, without generating them on the fly.

Reply:
I find that having the fully-expanded test cases makes them harder to keep track of because there are so many covering similar cases.

But we probably don't need all of the test cases duplicated pre/post-epoch, so I'll remove the post-epoch duplication and add a few more explicit tests instead.